File: composeApp/src/androidMain/kotlin/com/snapchef/app/features/auth/presentation/SignInViewModel.kt
package com.snapchef.app.features.auth.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.snapchef.app.core.auth.AuthManager
import com.snapchef.app.core.data.remote.createHttpClient
import com.snapchef.app.features.auth.data.remote.AuthApiService
import com.snapchef.app.features.auth.data.remote.LoginRequest
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class SignInUiState(
    val email: String = "",
    val password: String = "",
    val showPassword: Boolean = false,
    val isLoading: Boolean = false,
    val errorMessage: String? = null
)

class SignInViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(SignInUiState())
    val uiState: StateFlow<SignInUiState> = _uiState.asStateFlow()

    private val apiService = AuthApiService(createHttpClient())

    fun updateEmail(email: String) {
        _uiState.value = _uiState.value.copy(email = email, errorMessage = null)
    }

    fun updatePassword(password: String) {
        _uiState.value = _uiState.value.copy(password = password, errorMessage = null)
    }

    fun toggleShowPassword() {
        _uiState.value = _uiState.value.copy(showPassword = !_uiState.value.showPassword)
    }

    fun signIn(onSuccess: () -> Unit) {
        val email = _uiState.value.email.trim()
        val password = _uiState.value.password

        if (email.isBlank() || password.isBlank()) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please enter email and password.")
            return
        }

        val emailRegex = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$".toRegex()
        if (!email.matches(emailRegex)) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please enter a valid email address.")
            return
        }

        if (password.length < 8) {
            _uiState.value = _uiState.value.copy(errorMessage = "Password must be at least 8 characters long.")
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val response = apiService.login(LoginRequest(email, password))
                AuthManager.accessToken = response.accessToken
                AuthManager.currentUser = response.user
                onSuccess()
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "Login failed: Incorrect credentials or server error."
                )
            }
        }
    }
}

File: iosApp/iosApp/auth_screens/SignInViewModel.swift
//
//  SignInViewModel.swift
//  iosApp
//
//  Created by gergana on 3/28/26.
//

import Foundation
import Shared  

@MainActor
final class SignInViewModel: ObservableObject {

    @Published var email:         String  = ""
    @Published var password:      String  = ""
    @Published var showPassword:  Bool    = false
    @Published var isLoading:     Bool    = false
    @Published var errorMessage:  String? = nil

    private let apiService = SnapChefServiceLocator.shared.authApiService

    func updateEmail(_ value: String) {
        email        = value
        errorMessage = nil
    }

    func updatePassword(_ value: String) {
        password     = value
        errorMessage = nil
    }

    func toggleShowPassword() {
        showPassword.toggle()
    }

    func signIn(
        onSuccess:        @escaping () -> Void,
        onVerifyRequired: @escaping (String) -> Void
    ) {
        let trimmedEmail = email.trimmingCharacters(in: .whitespaces)

        guard !trimmedEmail.isEmpty, !password.isEmpty else {
            errorMessage = "Please enter both email and password"
            return
        }

        isLoading    = true
        errorMessage = nil

        Task {
            do {
                let response = try await apiService.login(
                    request: LoginRequest(email: trimmedEmail, password: password)
                )
                AuthManager.shared.accessToken = response.accessToken
                AuthManager.shared.currentUser = response.user
                onSuccess()
            } catch let error as KotlinException where isHttp403(error) {
                onVerifyRequired(trimmedEmail)
            } catch {
                isLoading    = false
                errorMessage = "Login failed. Check your credentials."
            }
        }
    }
    
    func googleSignIn(
        idToken:   String,
        onSuccess: @escaping () -> Void
    ) {
        isLoading    = true
        errorMessage = nil

        Task {
            do {
                let response = try await apiService.googleAuth(idToken: idToken)
                AuthManager.shared.accessToken = response.accessToken
                AuthManager.shared.currentUser = response.user
                onSuccess()
            } catch {
                isLoading    = false
                errorMessage = "Google Sign-In failed on the server."
            }
        }
    }

    private func isHttp403(_ error: KotlinException) -> Bool {
        return error.description().contains("403")
    }
}
